packages/core/core/src/AssetGraphBuilder.js

```diff
@@ -142,46 +136,162 @@ export default class AssetGraphBuilder extends EventEmitter {
-    } = this.graph.resolveTransformerRequest(req, cacheEntry);
-
-    if (this.watcher) {
-      for (let file of addedFiles) {
-        this.watcher.watch(file.filePath);
-      }
+    let cacheEntry = await Cache.get(cacheKey);
+    if (!cacheEntry) {
+      console.log('TRANSFORMING', req);
+
+      cacheEntry = await this.runTransform({
+        req,
+        configs,
+        devDeps
+      });
 
-      for (let file of removedFiles) {
-        this.watcher.unwatch(file.filePath);
+      let time = Date.now() - start;
+      for (let asset of cacheEntry.assets) {
+        asset.stats.time = time;
       }
     }
 
+    if (signal.aborted) throw new BuildAbortError();
+    let {newDepNodes} = this.graph.resolveTransformerRequest(req, cacheEntry);
+
     // The shallow option is used during the update phase
     if (!shallow) {
-      for (let dep of newDeps) {
-        this.queue.add(() => this.resolve(dep, {signal}));
+      for (let depNode of newDepNodes) {
+        this.queue.add(() => this.resolve(depNode, {signal}));
+      }
+    }
+  }
+
+  async loadBuildDependencies(node, buildOpts) {
+    let configRequest = {
+      filePath: node.value.filePath,
+      configType: 'parcel',
+      meta: {
+        actionType: node.type
       }
+    };
+
+    let configLoader = new ConfigLoader(this.options);
+
+    let {config, devDeps} = await this.loadConfigAndResolveDependencies(
+      configRequest,
+      configLoader,
+      node
+    );
+
+    let configs = {parcel: config};
+
+    return {configs, devDeps};
+
+    // await Promise.all(
+    //   devDeps.map(async devDep => {
+    //     let plugin = this.loadParcelPlugin(devDep.packageName, parcelConfig);
+    //     configRequest = plugin.getConfigRequest(node.value);
+    //     await this.loadConfigAndResolveDependencies(configRequest, node);
+    //   })
+    // );
+  }
+
+  async loadConfigAndResolveDependencies(
+    configRequest,
+    configLoader,
+    actionNode
+  ) {
+    let configRequestNode = nodeFromConfigRequest(configRequest);
+    let result;
+    let devDepRequestNodes;
+    if (
+      !this.graph.hasNode(configRequestNode) ||
+      this.graph.invalidNodes.has(configRequestNode.id)
+    ) {
+      this.graph.addConfigRequest(configRequestNode, actionNode);
+      console.log('LOADING CONFIG', configRequest);
+      result = await configLoader.load(configRequest);
+      let {devDepRequestNodes: ddrNodes} = this.graph.resolveConfigRequest(
+        result,
+        configRequestNode
+      );
+      devDepRequestNodes = ddrNodes;
+    } else {
+      devDepRequestNodes = this.graph.getConfigDevDepNodes(configRequestNode);
+    }
+
+    let devDeps = await Promise.all(
+      devDepRequestNodes.map(devDepRequestNode =>
+        this.resolveDevDep(devDepRequestNode, actionNode)
+      )
+    );
+
+    return {config: result.config, devDeps};
+  }
+
+  async resolveDevDep(devDepRequestNode, actionNode) {
+    let [devDepNode] = this.graph.getNodesConnectedFrom(
```

Comment:
Originally I was thinking the `AssetGraph.js` would be a pure data structure and `AssetGraphBuilder.js` would just execute the async functions and update the graph with the results, but now that the state of the graph is being used more to execute or not execute some async functions, the two are becoming more coupled. With the watcher code moved to `Parcel.js` I'm seeing less value in keeping them separate. When I change one, I often have to change the other. Thoughts on eventually merging these two files together?